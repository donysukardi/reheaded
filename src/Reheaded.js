import { Component } from 'react'
import PropTypes from 'prop-types'
import shallowequal from 'shallowequal'
import 'raf/polyfill';

import shouldUpdate from './shouldUpdate'

const noop = () => {}

class Reheaded extends Component {
  static propTypes = {
    parent: PropTypes.func,
    children: PropTypes.func.isRequired,
    disabled: PropTypes.bool,
    upTolerance: PropTypes.number,
    downTolerance: PropTypes.number,
    onPin: PropTypes.func,
    onUnpin: PropTypes.func,
    onUnfix: PropTypes.func,
    pinStart: PropTypes.number,
    calcHeightOnResize: PropTypes.bool,
  };

  static defaultProps = {
    parent: () => window,
    disabled: false,
    upTolerance: 5,
    downTolerance: 5,
    onPin: noop,
    onUnpin: noop,
    onUnfix: noop,
    pinStart: 0,
    calcHeightOnResize: true,
  };

  constructor (props) {
    super(props)
    // Class variables.
    this.currentScrollY = 0
    this.lastKnownScrollY = 0
    this.scrollTicking = false
    this.resizeTicking = false
    this.state = {
      state: 'unfixed',
      height: 0,
    }
  }

  componentDidMount () {
    const { disabled, parent: parentFn, calcHeightOnResize } = this.props;
    this.setHeightOffset();
    if (!disabled) {
      const parent = parentFn();
      parent.addEventListener('scroll', this.handleScroll)
      if (calcHeightOnResize) {
        parent.addEventListener('resize', this.handleResize)
      }
    }
  }

  shouldComponentUpdate (nextProps, nextState) {
    return (
      !shallowequal(this.props, nextProps) ||
      !shallowequal(this.state, nextState)
    )
  }

  componentDidUpdate (prevProps) {
    const parent = this.props.parent();
    if (this.props.disabled && !prevProps.disabled) {
      this.unfix()
      parent.removeEventListener('scroll', this.handleScroll)
      parent.removeEventListener('resize', this.handleResize)
    } else if (!this.props.disabled && prevProps.disabled) {
      parent.addEventListener('scroll', this.handleScroll);
      /* istanbul ignore else  */
      if (this.props.calcHeightOnResize) {
        parent.addEventListener('resize', this.handleResize);
      }
    }
  }

  componentWillUnmount () {
    const parent = this.props.parent();
    parent.removeEventListener('scroll', this.handleScroll);
    /* istanbul ignore else  */
    if(parent !== window) {
      window.removeEventListener('scroll', this.handleScroll);
    }
    parent.removeEventListener('resize', this.handleResize);
  }

  setRef = ref => {
    this.inner = ref;
  }

  setHeightOffset = () => {
    this.setState({
      height: this.inner.offsetHeight,
    })
    this.resizeTicking = false
  }

  getScrollY = () => {
    const parent = this.props.parent();
    /* istanbul ignore else  */
    if (parent.pageYOffset !== undefined) {
      return parent.pageYOffset
    } else if (parent.scrollTop !== undefined) {
      return parent.scrollTop
    }
    /* istanbul ignore next line */
    return (document.documentElement || document.body.parentNode || document.body).scrollTop
  }

  getViewportHeight = () => (
    /* istanbul ignore next line */
    window.innerHeight
      || document.documentElement.clientHeight
      || document.body.clientHeight
  )

  getDocumentHeight = /* istanbul ignore next */ () => {
    const { body, documentElement } = document;
    return Math.max(
      body.scrollHeight, documentElement.scrollHeight,
      body.offsetHeight, documentElement.offsetHeight,
      body.clientHeight, documentElement.clientHeight
    )
  }

  getElementPhysicalHeight = elm => Math.max(
    elm.offsetHeight,
    elm.clientHeight
  )

  getElementHeight = elm => Math.max(
    elm.scrollHeight,
    elm.offsetHeight,
    elm.clientHeight,
  )

  getScrollerPhysicalHeight = () => {
    const parent = this.props.parent()

    return (parent === window || parent === document.body)
      ? /* istanbul ignore next */ this.getViewportHeight()
      : this.getElementPhysicalHeight(parent)
  }

  getScrollerHeight = () => {
    const parent = this.props.parent()

    return (parent === window || parent === document.body)
      ? /* istanbul ignore next */ this.getDocumentHeight()
      : this.getElementHeight(parent)
  }

  isOutOfBound = (currentScrollY) => {
    const pastTop = currentScrollY < 0;

    const scrollerPhysicalHeight = this.getScrollerPhysicalHeight();
    const scrollerHeight = this.getScrollerHeight();

    const pastBottom = currentScrollY + scrollerPhysicalHeight > scrollerHeight;

    return pastTop || pastBottom;
  }

  handleScroll = () => {
    /* istanbul ignore else  */
    if (!this.scrollTicking) {
      this.scrollTicking = true
      window.requestAnimationFrame(this.update)
    }
  }

  handleResize = /* istanbul ignore next */ () => {
    /* istanbul ignore else  */
    if (!this.resizeTicking) {
      this.resizeTicking = true
      window.requestAnimationFrame(this.setHeightOffset)
    }
  }

  unpin = () => {
    this.props.onUnpin()
    this.setState({
      state: 'unpinned'
    })
  }

  pin = () => {
    this.props.onPin()
    this.setState({
      state: 'pinned',
    })
  }

  unfix = () => {
    this.props.onUnfix()
    this.setState({
      state: 'unfixed',
    })
  }

  update = () => {
    this.currentScrollY = this.getScrollY();
    /* istanbul ignore else  */
    if (!this.isOutOfBound(this.currentScrollY)) {
      const { action } = shouldUpdate(
        this.lastKnownScrollY,
        this.currentScrollY,
        this.props,
        this.state
      )

      /* istanbul ignore else  */
      if (action === 'pin') {
        this.pin()
      } else if (action === 'unpin') {
        this.unpin()
      } else if (action === 'unfix') {
        this.unfix()
      }
    }

    this.lastKnownScrollY = this.currentScrollY
    this.scrollTicking = false
  }

  render () {
    const { children } = this.props

    return (
      children({
        setRef: this.setRef,
        ...this.state,
      })
    )
  }
}

export default Reheaded;
