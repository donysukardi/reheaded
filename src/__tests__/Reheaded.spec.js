import React from 'react'
import PropTypes from 'prop-types'
import { cleanup, render } from 'react-testing-library'
import 'jest-dom/extend-expect'

import Reheaded from '../Reheaded'

beforeEach(() => {
  jest
    .spyOn(global.window, 'requestAnimationFrame')
    .mockImplementation(cb => cb())
})

afterEach(() => {
  cleanup()
  global.window.requestAnimationFrame.mockRestore()
})

const getParent = () => ({
  addEventListener: jest.fn(),
  removeEventListener: jest.fn(),
  pageYOffset: 0,
  scrollHeight: 1200,
  offsetHeight: 50,
})

const ReheadedExample = ({ reheadedRef, ...props }) => (
  <div style={{ minHeight: '100vh' }}>
    <Reheaded {...props} ref={reheadedRef}>
      {({ setRef, height, state }) => (
        <div
          data-testid="reheadedWrapper"
          style={{
            height,
          }}
        >
          <header ref={setRef} data-testid="headerContainer">
            <h1>This is header</h1>
            <div data-testid="ReheadedState">{state}</div>
          </header>
        </div>
      )}
    </Reheaded>
    <div>This is some content</div>
  </div>
)

ReheadedExample.propTypes = {
  reheadedRef: PropTypes.func,
}

test('Reheaded component renders properly', () => {
  const { getByTestId, container } = render(<ReheadedExample />)
  expect(getByTestId('ReheadedState')).toHaveTextContent('unfixed')
  expect(container.firstChild).toMatchSnapshot()
})

test('Reheaded does not register on resize event listener if calcHeightOnResize is false', () => {
  const parent = getParent()
  render(<ReheadedExample parent={() => parent} calcHeightOnResize={false} />)

  expect(parent.addEventListener.mock.calls[0][0]).toBe('scroll')
  expect(parent.addEventListener.mock.calls.length).toBe(1)
})

test('Reheaded didMount registers scroll event listener', () => {
  const parent = getParent()
  const { unmount } = render(<ReheadedExample parent={() => parent} />)

  expect(parent.addEventListener.mock.calls[0][0]).toBe('scroll')
  expect(parent.addEventListener.mock.calls[1][0]).toBe('resize')

  unmount()
  expect(parent.removeEventListener.mock.calls[0][0]).toBe('scroll')
  expect(parent.removeEventListener.mock.calls[1][0]).toBe('resize')
})

test('Reheaded toggling disabled on should remove event listeners', () => {
  const parent = getParent()
  const parentFn = () => parent

  const { rerender } = render(<ReheadedExample parent={parentFn} />)

  expect(parent.addEventListener.mock.calls[0][0]).toBe('scroll')
  expect(parent.addEventListener.mock.calls[1][0]).toBe('resize')

  rerender(<ReheadedExample parent={parentFn} disabled />)

  expect(parent.removeEventListener.mock.calls[0][0]).toBe('scroll')
  expect(parent.removeEventListener.mock.calls[1][0]).toBe('resize')
})

test('Reheaded toggling disabled off should add event listeners', () => {
  const parent = getParent()
  const parentFn = () => parent

  const { rerender } = render(<ReheadedExample parent={parentFn} disabled />)

  expect(parent.addEventListener).not.toHaveBeenCalled()

  rerender(<ReheadedExample parent={parentFn} />)

  expect(parent.addEventListener.mock.calls[0][0]).toBe('scroll')
  expect(parent.addEventListener.mock.calls[1][0]).toBe('resize')
})

test('Reheaded setHeightOffset updates the height in state', () => {
  const parent = getParent()
  let reheadedRef
  const setreheadedRef = ref => {
    reheadedRef = ref
  }

  const { getByTestId } = render(
    <ReheadedExample parent={() => parent} reheadedRef={setreheadedRef} />,
  )

  const height = '300px'
  reheadedRef.inner = {
    offsetHeight: height,
  }
  reheadedRef.setHeightOffset()
  expect(getByTestId('reheadedWrapper').style.height).toEqual(height)
})

test('Reheaded handle scroll', async () => {
  let reheadedRef
  const setreheadedRef = ref => {
    reheadedRef = ref
  }

  const onUnpinSpy = jest.fn()
  const onUnfixSpy = jest.fn()
  const onPinSpy = jest.fn()

  const parent = getParent()

  const { getByTestId } = render(
    <ReheadedExample
      onPin={onPinSpy}
      onUnpin={onUnpinSpy}
      onUnfix={onUnfixSpy}
      parent={() => parent}
      reheadedRef={setreheadedRef}
    />,
  )

  parent.pageYOffset = 300
  reheadedRef.handleScroll()

  expect(getByTestId('ReheadedState')).toHaveTextContent('unpin')
  expect(onUnpinSpy).toHaveBeenCalledTimes(1)

  parent.pageYOffset = 180
  reheadedRef.handleScroll()

  expect(getByTestId('ReheadedState')).toHaveTextContent('pinned')
  expect(onPinSpy).toHaveBeenCalledTimes(1)

  parent.pageYOffset = 0
  reheadedRef.handleScroll()

  expect(getByTestId('ReheadedState')).toHaveTextContent('unfixed')
  expect(onUnfixSpy).toHaveBeenCalledTimes(1)
})

test('Reheaded derives scrollY from scrollTop', async () => {
  let reheadedRef
  const setreheadedRef = ref => {
    reheadedRef = ref
  }

  const parent = getParent()
  delete parent.pageYOffset
  parent.scrollTop = 30

  render(<ReheadedExample parent={() => parent} reheadedRef={setreheadedRef} />)

  expect(reheadedRef.getScrollY()).toEqual(30)
})

test('Reheaded getViewportHeight from window.innerHeight', async () => {
  let reheadedRef
  const setreheadedRef = ref => {
    reheadedRef = ref
  }

  window.innerHeight = 1200

  render(<ReheadedExample reheadedRef={setreheadedRef} />)

  expect(reheadedRef.getViewportHeight()).toEqual(1200)
})
