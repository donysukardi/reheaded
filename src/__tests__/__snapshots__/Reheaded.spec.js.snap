// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reheaded component renders properly 1`] = `
<div
  style="min-height: 100vh;"
>
  <div
    data-testid="reheadedWrapper"
    style="height: 0px;"
  >
    <header
      data-testid="headerContainer"
    >
      <h1>
        This is header
      </h1>
      <div
        data-testid="ReheadedState"
      >
        unfixed
      </div>
    </header>
  </div>
  <div>
    This is some content
  </div>
</div>
`;
